internal/app/go-filecoin/paymentchannel/manager.go

```diff
@@ -236,6 +218,105 @@ func PaychActorCtorExecParamsFor(client, miner address.Address) (initActor.ExecP
+	view, err := pm.stateViewer.GetStateView(ctx, tok)
+	if err != nil {
+		return address.Undef, err
+	}
+	_, fcworker, err := view.MinerControlAddresses(ctx, miner)
+	return fcworker, err
+}
+
+func (pm *Manager) WaitForCreatePaychMessage(ctx context.Context, mcid cid.Cid) (address.Address, error) {
+	var newPaychAddr address.Address
+
+	handleResult := func(b *block.Block, sm *types.SignedMessage, mr *vm.MessageReceipt) error {
+		var res initActor.ExecReturn
+		if err := encoding.Decode(mr.ReturnValue, &res); err != nil {
+			return err
+		}
+
+		var msgParams initActor.ExecParams
+		if err := encoding.Decode(sm.Message.Params, &msgParams); err != nil {
+			return err
+		}
+
+		var ctorParams *paychActor.ConstructorParams
+		if err := encoding.Decode(msgParams.ConstructorParams, &ctorParams); err != nil {
+			return err
+		}
+
+		newPaychAddr = res.RobustAddress
+		return nil
+	}
+
+	err := pm.waiter.Wait(pm.ctx, mcid, handleResult)
+	if err != nil {
+		return address.Undef, err
+	}
+	return newPaychAddr, nil
+}
+
+func (pm *Manager) AddFundsToChannel(paychAddr address.Address, amt abi.TokenAmount) (cid.Cid, error) {
+	var chinfo ChannelInfo
+	st := pm.paymentChannels.Get(paychAddr)
+	if err := st.Get(&chinfo); err != nil {
+		return cid.Undef, err
+	}
+
+	mcid, _, err := pm.sender.Send(context.TODO(), chinfo.From, paychAddr, amt, defaultGasPrice, defaultGasLimit, true, builtin.MethodSend, nil)
+	if err != nil {
+		return cid.Undef, err
+	}
+	// TODO: track amts in paych store
+	return mcid, nil
+}
+
+func (pm *Manager) WaitForAddFundsMessage(ctx context.Context, mcid cid.Cid) error {
+	handleResult := func(b *block.Block, sm *types.SignedMessage, mr *vm.MessageReceipt) error {
+		if mr.ExitCode != exitcode.Ok {
+			return xerrors.Errorf("Add funds failed with exitcode %d", mr.ExitCode)
+		}
+		return nil
+	}
+	return pm.waiter.Wait(pm.ctx, mcid, handleResult)
+}
+
+// WaitForPaychCreateMsg waits for mcid to appear on chain and returns the robust address of the
+// created payment channel
+// TODO: wait outside store lock, also set up channel tracking somehow
+// before knowing paych addr
+func (pm *Manager) handlePaychCreateResult(ctx context.Context, mcid cid.Cid, client, miner address.Address) {
+
+	handleResult := func(_ *block.Block, _ *types.SignedMessage, mr *vm.MessageReceipt) error {
+		if mr.ExitCode != 0 {
+			log.Errorf("create message failed with exit code %d", mr.ExitCode)
+		}
+
+		var decodedReturn initActor.ExecReturn
+		if err := decodedReturn.UnmarshalCBOR(bytes.NewReader(mr.ReturnValue)); err != nil {
+			return err
+		}
+		paychAddr := decodedReturn.RobustAddress
+
+		// TODO check again to make sure a payment channel has not been created for this From/To
+		chinfo := ChannelInfo{
+			From:       client,
+			To:         miner,
+			NextLane:   0,
+			NextNonce:  1,
+			UniqueAddr: paychAddr,
+		}
+		return pm.paymentChannels.Begin(paychAddr, &chinfo)
```

Comment:
This really should be moved outside the Wait callback. The callback is just to control access to message state.